source/opt/aggressive_dead_code_elim_pass.cpp

```diff
@@ -438,11 +438,12 @@ std::vector<uint32_t> AggressiveDCEPass::GetLoadedVariablesFromFunctionCall(
     const Instruction* inst) {
   assert(inst->opcode() == spv::Op::OpFunctionCall);
   std::vector<uint32_t> live_variables;
-  inst->ForEachInId([this, &live_variables](const uint32_t* operand_id) {
-    if (!IsPtr(*operand_id)) return;
-    uint32_t var_id = GetVariableId(*operand_id);
-    live_variables.push_back(var_id);
-  });
+  inst->ForEachFunctionCallParamId(
```

Comment:
The intent here is to just skip the first operand of `OpFunctionCall`. I'm not sure if this is the correct approach/style for doing that.

Reply:
I don't think this is the right solution. The problem is that `IsPtr` returns the wrong result when it is given the `OpFunction`. The SPIR-V is ackward, but the "result type" of an OpFunction is the return type of the function. `IsPtr` sees that the result type of the id is a pointer, and return true. That is incorrect. It should have another check that the opcode is not OpFunction. Then this part of the code can remain the same.

Reply:
@s-perron Can you please explain why this is not the correct general solution i.e. skipping the first operand to the OpFunctionCall? If you look at the name of this function (GetLoadedVariablesFromFunctionCall), it is intended that only the actual arguments to the call are to be considered. The first operand to the OpFunctionCall instruction ie the OpFunction should not be looked at here at all. Only the 2nd operand and beyond should be considered. This is a day 1 bug (not skipping the OpFunction operand) that we have only avoided because it is rare to see a function call in ADCE, let alone one that return a pointer type.

As far as I can tell, IsPtr is not intended to take an OpFunction instruction, so I am not understanding why we want to make that work. Can you please explain why IsPtr should be changed handle an OpFunction?

Reply:
> Can you please explain why IsPtr should be changed handle an OpFunction?

I ask the question "What should `IsPtr` return when given the id of an `OpFunction`. The things I consider are:

1.  Returning false for `OpFunction` ids will be consistent with what a developer would expect the function to do. The name is `IsPtr` and the `OpFunction` id is obviously not a pointer. Also, the comments at the function declaration describe what the function should return:

https://github.com/KhronosGroup/SPIRV-Tools/blob/29431859f575633790365a0ac841b27440274f42/source/opt/mem_pass.h#L85-L87

An `OpFunction` is none of the three criteria, so `IsPtr` should return false when given the id of an `OpFunction`.

2. Returning false all of the time will be consistent with previous behaviour. As you mentioned, the function has always returned false for `OpFunction` ids until recently.
3. Saying that `IsPtr` is undefined when it is given the id of an `OpFunction` instruction, forces developers to do something to make sure that the input to `IsPtr` is not an `OpFunction`. This could be easily forgotten because it is not intuitive, and could lead to more errors.

>  As far as I can tell, IsPtr is not intended to take an OpFunction instruction

I see nothing on the code to justify this statement. I think it was more likely that the case was not considered as a special case.

Reply:
Agreed.  If `IsPtr` receives a non-pointer id it should trivially return `false`.

Reply:
It seems like there are 2 issues here:
1. `IsPtr` does not correctly handle `OpFunction` as input
2. The `ForEachInId` loop in `GetLoadedVariablesFromFunctionCall` should only be iterating over function call parameters, not the `OpFunction` id.

Resolving either one will fix the the original issue. If you only want (1) resolved, I can do that, but then it seems like a comment is warranted in `GetLoadedVariablesFromFunctionCall` making it clear the `OpFunction` id will be processed, but skipped. I could also see keeping code similar to what I have (again, not sure if it's "correct" style-wise, though) and also fixup `IsPtr` to handle `OpFunction`.

Reply:
`IsPtr` has to be fixed. You can add a comment for 2. That should be enough. It is not wrong to include the function id as one iteration for the loop. The problem comes in with what you do with it, and that is correct if `IsPtr` is fixed up.